inject/src/main/java/io/micronaut/inject/annotation/EvaluatedAnnotationMetadata.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.inject.annotation;
+
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.BeanDefinitionAware;
+import io.micronaut.context.ContextConfigurable;
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.inject.BeanDefinition;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * Variation of {@link AnnotationMetadata} that is used when evaluated expression
+ * in annotation values need to be resolved at runtime.
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0
+ */
+@Internal
+public final class EvaluatedAnnotationMetadata extends MappingAnnotationMetadataDelegate implements ContextConfigurable, BeanDefinitionAware {
```

Comment:
I think we can simplify by defining it as `class EvaluatedAnnotationMetadata extends DefaultAnnotationMetadata` and if the evaluation is required just make a copy and use `EvaluatedAnnotationMetadata` with overridden `newAnnotationValue`

Reply:
It seems like `DefaultAnnotationMetadata` is not really designed to be easily subclassed. Its set of constructors demands map of annotation values which seem to be not so easy to obtain from other instances of `AnnotationMetadata`. This `EvaluatedAnnotationMetadata` is designed as a wrapper aware of expressions which can wrap any `AnnotationMetadata` not only `DefaultAnnotationMetadata`